### PR TITLE
Use embedded zxing lib in BarcodeWidget. Fixes #1021

### LINKED ISSUE #1021: Use embedded zxing lib in BarcodeWidget
#843 introduced the zxing library for reading barcodes because it's a lot more convenient for users who want to use the QR code import feature. Now that we have the dependency, the BarcodeWidget should use it instead of requiring users to install the ZXing app.

### PR DESCRIPTION
**Pull request by MukundAnanthu**
Pull request for issue 1021. Uses embedded zxing lib in BarcodeWidget instead of launching intent for an external Barcode scanning app.

### CODE REVIEW

**Comment by shobhitagarwal1612** [comment]
Hey @MukundAnanthu, 
IntentIntegrator only sets up an intent for scanning a barcode. We have to get the results by overriding the onActivityResult method. 
Here, [onActivityResult](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L572) is already overridden in FormEntryActivity. So, you just need to replace the existing code for fetching the results in that method.

```
// sample code
@Override
protected void onActivityResult(int requestCode, int resultCode, Intent data) {
    IntentResult result = IntentIntegrator.parseActivityResult(requestCode, resultCode, data);
    if(result != null) {
        if(result.getContents() == null) {
            // scanning cancelled
        } else {
            String data = result.getContents();
        }
    } else {
        super.onActivityResult(requestCode, resultCode, data);
    }
}
```

Link to [zxing library](https://github.com/journeyapps/zxing-android-embedded#usage-with-intentintegrator)

P.S. There is a custom scanner activity that was added for #781 (See usage [here](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/fragments/ShowQRCodeFragment.java#L136)) which is customised and also has a flashlight button. You can use the same class here too.

**Comment by MukundAnanthu** [comment]
Thanks @shobhitagarwal1612 for the review. Have made some more changes, would love to hear your feedback

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 1]
```diff
@@ -595,12 +598,23 @@ protected void onActivityResult(int requestCode, int resultCode,
...
-        switch (requestCode) {
-            case BARCODE_CAPTURE:
+        // For handling results returned by the Zxing Barcode scanning library
+        IntentResult barcodeScannerResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, intent);
+        if (barcodeScannerResult != null) {
+            if (barcodeScannerResult.getContents() == null) {
+                // request was canceled...
+                Timber.i("QR code scanning cancelled");
+            } else {
                 String sb = intent.getStringExtra("SCAN_RESULT");
```
Comment (shobhitagarwal1612) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:608]: This should be replaced with the following line, 
`String sb = barcodeScannerResult.getContents();`

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 2]
```diff
@@ -595,12 +598,23 @@ protected void onActivityResult(int requestCode, int resultCode,
...
                 ((ODKView) currentView).setBinaryData(sb);
                 saveAnswersForCurrentScreen(DO_NOT_EVALUATE_CONSTRAINTS);
-                break;
+                return;
```
Comment (shobhitagarwal1612) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: This statement makes the `refreshCurrentView()` line at the end of this method ([line 761](https://github.com/MukundAnanthu/collect/blob/27a051c1bf64ea38a0b001b6e2eace8c7c378ab7/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L761)) unreachable.
Please add this line `refreshCurrentView()` before returning from this method.

**Review by shobhitagarwal1612 — CHANGES_REQUESTED** [review]

File: collect_app/src/main/java/org/odk/collect/android/widgets/BarcodeWidget.java [hunk 3]
```diff
@@ -68,21 +66,10 @@ public void onClick(View v) {
...
-                try {
-                    Collect.getInstance().getFormController()
-                            .setIndexWaitingForData(formEntryPrompt.getIndex());
-                    ((Activity) getContext()).startActivityForResult(i,
-                            FormEntryActivity.BARCODE_CAPTURE);
-                } catch (ActivityNotFoundException e) {
-                    Toast.makeText(
-                            getContext(),
-                            getContext().getString(
-                                    R.string.barcode_scanner_error),
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/widgets/BarcodeWidget.java:81]: Please remove this string from strings.xml file as it's no longer needed.

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java (hunk 1 above)
Comment (MukundAnanthu) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:608]: @shobhitagarwal1612 I actually wrote both statements in the same place just for comparison..i.e 
Line 1: String sb1 = barcodeScannerResult.getContents();
and 
Line 2: String sb2 = intent.getStringExtra("SCAN_RESULT");

Turns out the value stored in sb1 and sb2 are exactly identical.i.e both statements are storing the same value in sb. 

Would you still recommend replacing the current statement
String sb = intent.getStringExtra("SCAN_RESULT"); 
with
 String sb = barcodeScannerResult.getContents();
If yes, why?

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java (hunk 2 above)
Comment (MukundAnanthu) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: @shobhitagarwal1612 Yes, will make this change to ensure refreshCurrentView() is reachable

File: collect_app/src/main/java/org/odk/collect/android/widgets/BarcodeWidget.java (hunk 3 above)
Comment (MukundAnanthu) [reply collect_app/src/main/java/org/odk/collect/android/widgets/BarcodeWidget.java:81]: @grzesiek2010 Sure!!

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java (hunk 1 above)
Comment (shobhitagarwal1612) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:608]: I am not sure if the later one will work as I used the way it was mentioned in the zxing library documentation. But if it's working then I guess it doesn't matter which one is used.

**Comment by lognaturel** [comment]
Thank you so much, @MukundAnanthu! This is a way better user experience and hopefully will encourage more organizations to use the power of barcodes.

And fantastic teamwork with @shobhitagarwal1612 and @grzesiek2010 as well! Go team. 😎

**Pull request merged**